src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -1632,4 +1645,217 @@ private static unsafe IntPtr ObjectToComWeakRef(object target, out long wrapperI
+                }
+
+                // _numEntries is only maintained for the purposes of deciding whether to
+                // expand the bucket and is not used during iteration to handle the
+                // scenario where element is in bucket but _numEntries hasn't been incremented
+                // yet.
+                _numEntries++;
+                if (_numEntries > (_buckets.Length * 2))
+                {
+                    ExpandBuckets();
+                }
+            }
+        }
+
+        private void ExpandBuckets()
+        {
+            try
+            {
+                int newNumBuckets = _buckets.Length * 2 + 1;
+                Entry?[] newBuckets = new Entry[newNumBuckets];
+                for (int i = 0; i < _buckets.Length; i++)
+                {
+                    Entry? entry = _buckets[i];
+                    while (entry != null)
+                    {
+                        Entry? nextEntry = entry.m_next;
+
+                        int bucket = GetBucket(entry.m_value, newNumBuckets);
+
+                        // We are allocating new entries for the bucket to ensure that
+                        // if there is an enumeration already in progress, we don't
+                        // modify what it observes by changing next in existing instances.
+                        Entry newEntry = new Entry()
+                        {
+                            m_value = entry.m_value,
+                            m_next = newBuckets[bucket],
+                        };
+                        newBuckets[bucket] = newEntry;
+
+                        entry = nextEntry;
+                    }
+                }
+                _buckets = newBuckets;
+            }
+            catch (OutOfMemoryException)
+            {
+            }
+        }
+
+        public void Remove(GCHandle handle)
+        {
+            using (_lock.EnterScope())
+            {
+                int bucket = GetBucket(handle);
+                Entry? prev = null;
+                Entry? entry = _buckets[bucket];
+                while (entry != null)
+                {
+                    if (handle.Equals(entry.m_value))
+                    {
+                        if (prev == null)
+                        {
+                            _buckets[bucket] = entry.m_next;
+                        }
+                        else
+                        {
+                            prev.m_next = entry.m_next;
+                        }
+                        _numEntries--;
+                        return;
+                    }
+
+                    prev = entry;
+                    entry = entry.m_next;
+                }
+            }
+        }
+
+        private int GetBucket(GCHandle handle, int numBuckets = 0)
+        {
+            int h = handle.GetHashCode();
+            h &= 0x7fffffff;
+            return (h % (numBuckets == 0 ? _buckets.Length : numBuckets));
```

Comment:
```suggestion
        private static int GetBucket(GCHandle handle, int numBuckets)
        {
            int h = handle.GetHashCode();
            return (int)((uint)h % (uint)numBuckets);
```
Passing numBuckets explicitly would make this easier to understand.
Also, unsigned division should be a bit more efficient.